src/Text/Pandoc/Readers/Vimwiki.hs

```diff
@@ -0,0 +1,398 @@
+import Text.Pandoc.Parsing (readWithM, ParserT, stateOptions, ParserState, blanklines, registerHeader, spaceChar, stateAllowLinks, emailAddress, guardEnabled, uri)
+import Text.Parsec.Char (spaces, char, anyChar, newline, string, noneOf)
+import Text.Parsec.Error (ParseError)
+import Text.Parsec.Combinator (eof, choice, many1, manyTill, count, skipMany1, notFollowedBy)
+import Text.Parsec.Pos (sourceColumn)
+import Text.Parsec.Prim (many, getPosition, try, runParserT)
+import Text.Parsec.String (Parser)
+import Text.Parsec (parse)
+import Text.Parsec.Char (oneOf, space)
+import Text.Parsec.Combinator (lookAhead, between)
+import Text.Parsec.Prim ((<|>), (<?>), skipMany)
+import Text.Pandoc.Options (Extension(Ext_autolink_bare_uris))
+
+readVimwiki :: PandocMonad m => ReaderOptions -> String -> m Pandoc
+readVimwiki opts s = do
+  res <- readWithM parseVimwiki def{ stateOptions = opts } s
+  case res of
+       Left e -> throwError e
+       Right result -> return result
+
+type VwParser = ParserT [Char] ParserState
+
+
+-- constants
+
+specialChars :: [Char]
+specialChars = "=*-#[]_~{`$|:%"
+
+spaceChars :: [Char] 
+spaceChars = " \t\n"
+
+-- main parser
+
+parseVimwiki :: PandocMonad m => VwParser m Pandoc
+parseVimwiki = do
+  bs <- mconcat <$> many block
+  spaces
+  eof
+  return $ B.doc bs
+
+-- block parser
+
+block :: PandocMonad m => VwParser m Blocks
+block = do
+  pos <- getPosition
+  res <- choice [ mempty <$ blanklines
+                , header
+                , hrule
+                , mempty <$ comment
+                , mixedList
+                , blockQuote
+                , preformatted
+                , displayMath 
+                , table
+                , para
+                ]
+  report $ ParsingTrace (take 60 $ show $ B.toList res) pos
+  return res
+
+header :: PandocMonad m => VwParser m Blocks
+hrule :: PandocMonad m => VwParser m Blocks
+displayMath :: PandocMonad m => VwParser m Blocks
+para :: PandocMonad m => VwParser m Blocks
+mixedList :: PandocMonad m => VwParser m Blocks
+table :: PandocMonad m => VwParser m Blocks
+blockQuote :: PandocMonad m => VwParser m Blocks
+preformatted :: PandocMonad m => VwParser m Blocks
+
+header = try $ do
+  sp <- many spaceChar 
+  eqs <- many1 (char '=')
+  spaceChar
+  let lev = length eqs
+  guard $ lev <= 6
+  contents <- trimInlines . mconcat <$> manyTill inline (try $ spaceChar >> (string eqs) >> many spaceChar >> (char '\n')) 
+  attr <- registerHeader ("", (if sp == "" then [] else ["justcenter"]), []) contents
+  return $ B.headerWith attr lev contents
+para = try $ do
+  contents <- trimInlines . mconcat <$> many1 inline
+  if all (==Space) contents
+     then return mempty
+     else return $ B.para contents
+hrule = try $ do
+  string "----" >> many (char '-') >> newline
+  return B.horizontalRule
+
+comment :: PandocMonad m => VwParser m ()
+comment = try $ do
+  many spaceChar >> string "%%" >> many (noneOf "\n") >> (lookAhead newline)
+  return ()
```

Comment:
Might be able to simplify such that comment is an line instead of a block level element. This might allow you to delegate matching leading whitespace to the whitespace inline parser. Just guessing, not certain yet.